pkg/testutils/api.go

```diff
@@ -0,0 +1,56 @@
+package testutils
+
+import (
+	"context"
+	"database/sql"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/pingcap/log"
+	"github.com/stretchr/testify/require"
+	"github.com/xmtp/xmtpd/pkg/api"
+	"github.com/xmtp/xmtpd/pkg/db/queries"
+	mocks "github.com/xmtp/xmtpd/pkg/mocks/registry"
+	"github.com/xmtp/xmtpd/pkg/proto/xmtpv4/message_api"
+	"github.com/xmtp/xmtpd/pkg/registrant"
+	"github.com/xmtp/xmtpd/pkg/registry"
+)
+
+func NewTestAPIServer(t *testing.T) (*api.ApiServer, *sql.DB, func()) {
+	ctx, cancel := context.WithCancel(context.Background())
+	log := NewLog(t)
+	db, _, dbCleanup := NewDB(t, ctx)
+	privKey, err := crypto.GenerateKey()
+	require.NoError(t, err)
+	privKeyStr := "0x" + HexEncode(crypto.FromECDSA(privKey))
+	mockRegistry := mocks.NewMockNodeRegistry(t)
+	mockRegistry.EXPECT().GetNodes().Return([]registry.Node{
+		{NodeID: 1, SigningKey: &privKey.PublicKey},
+	}, nil)
+	registrant, err := registrant.NewRegistrant(ctx, queries.New(db), mockRegistry, privKeyStr)
+	require.NoError(t, err)
+
+	svr, err := api.NewAPIServer(ctx, db, log, 0 /*port*/, registrant, true /*enableReflection*/)
+	require.NoError(t, err)
+
+	return svr, db, func() {
+		log.Info("-------- Cleaning up server and DB ----------")
+		cancel()
```

Comment:
By canceling the context, we clean up dependencies such as the subscribe and publish worker, so that they will stop querying the DB and returning errors.

I'm not sure if going forward, we should keep public `Close()` methods, or rely on cleanup being triggered via closing the context.

Reply:
If the cleanup is just "stop this thing" it's fine (and preferable) to do it through the context. 

The one advantage to having a close method is waiting for cleanup tasks that take time. Calling `Close` as part of shutdown can make sure that the `Close` method has completed before the server fully terminates. The server may shut down before all listeners to the `context.Done()` channel have picked up the message.